app/models/question_mover.rb

```diff
@@ -0,0 +1,79 @@
+# Moves questions within a QuestionSet's question tree.
+# Checks to make sure the move is valid before executing it.
+class QuestionMover
+  include ActiveModel::Model
+
+  attr_accessor :current_division, :question, :target, :relation
+
+  def move
+    raise ArgumentError, "invalid relation" unless %i[before after inside].include?(relation)
+
+    ensure_current_division
+    ensure_new_parent_is_group
+    ensure_correct_adjacency
+    ensure_parent_of_same_or_ancestor_division
+    method =
+      case relation
+      when :before then :prepend_sibling
+      when :after then :append_sibling
+      when :inside then :prepend_child
+      end
```

Comment:
does prepending for inside here mean that as a user, if i have group A with qs A1, A2, and A4, and I want to pull in an existing q A3 from elsewhere, A3 will first be added to A, and they'll be A3, A1, A2, A4. Then I can move A3 after A2 in a second step? that's ok, just seeing if i understand this right.

Reply:
Good question! This comes from the jqtree library. It will let you drag A3 into position directly and send it to the server as `:after`. I think `:inside` is only ever used if the group is empty.